Preliminary Support for `ffcv`

rahul-tuli opened the pull request:
`FFCV` or `Fast Forward Computer Vision` is a library that claims upto 7x speedup over vanilla `pytorch` for training flows. This PR provides preliminary support for FFCV with `sparseml` for `ImageNet` based `image_classification` models.

To apply FFCV to your existing training flow simply append `--ffcv` flag when invoking the `sparseml.image_classification.train` command line utility.

Currently supported Datasets are:
- [X] ImageNet
- [X] Imagenette

Some other things to note
- [X] FFCV installation pathway is as follows
```bash
# ffcv installation
conda create -y -n ffcv python=3.9 cupy pkg-config compilers libjpeg-turbo opencv pytorch torchvision cudatoolkit=11.3 numba -c pytorch -c conda-forge
pip install ffcv

# clone and install sparseml from source
git clone git@github.com:neuralmagic/sparseml
pip install -e "./sparseml/[dev,torchvision]"
``` 

- [X] The test command is as follows
```
sparseml.image_classification.train --recipe-path "zoo:cv/classification/resnet_v1-50/pytorch/sparseml/imagenet/pruned95-none?recipe_type=original" --arch-key resnet50 --checkpoint-path zoo --dataset imagenette --train-batch-size 128 --test-batch-size 256 --dataset-path dataset --loader-num-workers 8 --ffcv
```

- [X] ~~Need to figure out a graceful exit strategy if `FFCV` is not installed and making `FFCVImageNetDataset` a co-parent will not break anything~~
- [X] Current PR relies on the new Image classification trainer, and helpers refactor #666  and in it's current state contains commits from that PR

Files to be reviewed from this PR are:
- imagenet.py
- imagenette.py
- ffcv_dataset.py
- helpers.py `get_dataset_and_dataloader(...)` function
- train.py

- [X] Docs 
- [X] change base branch to main before merging

rahul-tuli committed:
Refactor:`sparseml.pytorch.datasets.classification.ffcv` package into  `sparseml/pytorch/datasets/ffcv_dataset.py`
Add: 'ffcv' support to trainer
Add: Helper `FFCVDataset` and `ImageFolderFFCV` classes

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+        """
+        Write the dataset to disk.
+        """
+        pass
```
nit: `raise NotImplementedError()` instead of pass

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+        """
+        Switch off the transform.
+        """
+        pass
```
nit: `raise NotImplementedError()` instead of pass

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
```
let's move this out of `/datasets` into `/datasets`

bfineran replied:
let's rename file to `ffcv_dataset.py`

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+        writer.from_indexed_dataset(self.dataset)
+
+    @contextmanager
+    def switched_off_transform(self):
```
let's move all this into write in the parent class since write will be a concrete method

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+
+        try:
+            # Switch off the transforms
+            if hasattr(self.dataset, "transform"):
```
there is also a property called `transforms` in Dataset and you'll need to account for this as well

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+
+    def get_loader(
+        self,
+        num_workers: int = 16,
+        batch_size: int = 32,
+        distributed: int = 0,
+        in_memory: int = 0,
+        validation: bool = False,
```
let's not default at least num workers and batch size, use judgement on the rest

rahul-tuli replied:
These are overwritten when called from the train script, but would like to sync and figure out the right way forward

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+        self,
+        num_workers: int = 16,
+        batch_size: int = 32,
+        distributed: int = 0,
+        in_memory: int = 0,
```
should make these bools and translate as needed. want to avoid having int inputs that need to be 1/0

rahul-tuli replied:
Done!

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+        :param batch_size: Batch size for the data loader.
+        :param distributed: 1/0 Whether to use distributed data loading.
+        :param in_memory: 1/0 Does the dataset fit in memory.
+        """
```
missing doc for validation

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+        LOGGER.info(_init_msg)
+
+        resolution = (self._image_size, self._image_size)
+        decoder = (
```
need to extensively check that for our image classification pipelines these are equivalent to our current transforms

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+            ToTensor(),
+            ToDevice(torch.device(self._device), non_blocking=True),
+            ToTorchImage(),
+            NormalizeImage(self.IMAGENET_MEAN, self.IMAGENET_STD, np.float32),
```
might want to have an overridable property for mean/std since some datasets might want to use a different one

rahul-tuli replied:
Let's sync on this offline

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+        Path(self.write_path).parent.mkdir(parents=True, exist_ok=True)
+
+    @abstractmethod
+    def _write(self, *args, **kwargs):
```
since the main logic for write is shared across ffcv (as far as we know), let's make this concrete and an abstract class to determine the fields. (make a subclass for image datasets that makes this concrete as well)

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+        """
+        self.dataset = dataset
+        self.write_path = write_path
+        Path(self.write_path).parent.mkdir(parents=True, exist_ok=True)
```
our dataset classes already have default and overridable locations for datasets to read from. Can probably just add an `default_dataset_path` folder to that path

bfineran commented on src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py:
```diff
@@ -0,0 +1,234 @@
+        pass
+
+
+class ImageFolderFFCV(FFCVDataset):
```
considering restructuring instead of a wrapper class to make it a co-parent class for our datasets - this will give each dataset the potential to override transforms used

bfineran reviewed (changes requested):
High level refactor:

Make the FFCVDataset be a co-parent class for our existing dataset classes

Looks like we need two main abstract methods (which can be defaulted with an intermediate parent class) (1) get the dataset fields (ie image, label) (2) get the transforms

Concrete methods can be made for write and get data loader

i.e.
FFCVDataset(abstractclass)
ImageFolderFFCVDataset(FFCVDataset) (ie would make both fields and transformations concerete)
....

ImageNetDataset(Dataset, ImageFolderFFCVDataset)

rahul-tuli committed:
Refactor: The design and add functionality

rahul-tuli committed:
Addressed Review Comments

* Made `FFCVImageNetDataset` a co-parent for ImageNet and Imagenette

* Created an abstract Interface to serve as a contract for FFCV data-writing and loading `FFCVCompatibleDataset`

bfineran commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,287 @@
+        :param in_memory: bool Does the dataset fit in memory.
+        :param device: str or int The device to use for the data loader.
+        """
+
```
let's raise an exception here if ffcv not available

rahul-tuli committed:
Fix: Graceful exit when FFCV not installed
Raise: A UserWarning noting the feature is experimental
Change: FFCV Type Annotations to be strings to postpone failure if FFCV not installed.

rahul-tuli committed:
Add: `__all__` to `ffcv_dataset.py`

bfineran commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,299 @@
+            return
+
+        # Make parents if they don't exist
+        dataset_path.parent.mkdir(parents=True, exist_ok=True)
```
any reason not to use `create_dirs` helper method here?

bfineran commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,299 @@
+            if hasattr(self, arg):
+                old_arg_value = getattr(self, arg)
+                setattr(self, arg, None)
+                LOGGER.info(f"Dataset attribute {arg} set to {None}")
```
let's make this a debug log

rahul-tuli replied:
Good idea! Thanks

bfineran commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,299 @@
+            # Restore dataset attribute to its original state
+            if old_arg_value is not None and hasattr(self, arg):
+                setattr(self, arg, old_arg_value)
+                LOGGER.info(f"Dataset attribute {arg} restored to old state")
```
let's make this a debug log

bfineran commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,299 @@
+        device: Union[str, int] = default_device(),
+    ):
+        """
+        Initialize the data loader.
```
specify ffcv

bfineran commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,299 @@
+        :param distributed: bool Whether to use distributed data loading.
+        :param in_memory: bool Does the dataset fit in memory.
+        :param device: str or int The device to use for the data loader.
+        """
```
nit: missing return comment

bfineran commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,299 @@
+            ffcv_url = "https://github.com/libffcv/ffcv#readme"
+            raise ImportError(
+                "FFCV is not installed. Please install using instructions from "
+                f"{ffcv_url}"
```
let's not point to the ffcv url for now, as we may add our own docs. can just say FFCV not installed

bfineran commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,299 @@
+    DEFAULT_CROP_RATIO = 224 / 256
+
+    @property
+    def ffcv_fields(self) -> "Dict[str, Field]":
```
```suggestion
    def ffcv_fields(self) -> Dict[str, "Field"]:
```

bfineran commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,299 @@
+    def ffcv_pipelines(
+        self,
+        device: Union[str, int] = default_device(),
+    ) -> "Dict[str, List[Operation]]":
```
```suggestion
    ) -> Dict[str, List["Operation"]]:
```

bfineran commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,299 @@
+        return loader
+
+
+class FFCVImageNetDataset(FFCVCompatibleDataset):
```
let's move this class to a file in the `pytorch/datasets/image_classification` directory since it is specific to that domain

bfineran commented on src/sparseml/pytorch/image_classification/utils/helpers.py:
```diff
@@ -214,6 +217,53 @@ def _create_val_dataset_and_loader(
+        isinstance(train_dataset, FFCVCompatibleDataset)
+        and isinstance(val_dataset, FFCVCompatibleDataset)
+    ):
+        raise ValueError("The Datasets have to be FFCV compatible for using ffcv")
```
```suggestion
        raise ValueError(
            f"datasets {type(train_dataset)} and {type(val_dataset)} must "
            "implement FFCVCompatibleDataset"
        )
```

bfineran commented on src/sparseml/pytorch/image_classification/utils/helpers.py:
```diff
@@ -214,6 +217,53 @@ def _create_val_dataset_and_loader(
+        raise ValueError("The Datasets have to be FFCV compatible for using ffcv")
+
+    train_loader = train_dataset.get_ffcv_loader(
+        write_path="data/ffcv-train.beton",
```
hardcoded value here

rahul-tuli replied:
defaulted to using `default_dataset_path(f"{args.dataset}_ffcv_train.beton")`

bfineran commented on src/sparseml/pytorch/image_classification/utils/helpers.py:
```diff
@@ -214,6 +217,53 @@ def _create_val_dataset_and_loader(
+    )
+
+    val_loader = val_dataset.get_ffcv_loader(
+        write_path="data/ffcv-val.beton",
```
hardcoded value here

rahul-tuli replied:
Defaulted to `default_dataset_path(f"{args.dataset}_ffcv_val.beton")`

bfineran reviewed (changes requested):
great first pass @rahul-tuli!

rahul-tuli commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,299 @@
+LOGGER = logging.getLogger(__name__)
+
+
+def timeit(func):
```
It can be removed But Noting I think logging the time taken to write an FFCVDataset to disk is good information for the users.

rahul-tuli commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,299 @@
+    @abstractmethod
+    def validation(self):
+        """
+        :returns The dataset to use
```
Good catch thanks!

rahul-tuli commented on src/sparseml/pytorch/datasets/ffcv_dataset.py:
```diff
@@ -0,0 +1,299 @@
+
+class FFCVCompatibleDataset(ABC):
+    """
+    Contract for adding FFCV functionality to a dataset.
```
Thanks will do!

rahul-tuli committed:
Initial Commit to reduce coupling b/w helpers and `TrainingArgs`

rahul-tuli committed:
Initial Commit to reduce coupling b/w helpers and `TrainingArgs`

rahul-tuli committed:
Updated signature + usages of `helpers.get_save_dir_and_loggers`

rahul-tuli committed:
Updated signature + usages of `helpers.get_save_dir_and_loggers`

rahul-tuli committed:
Updated signature + usages of `helpers.create_model`

rahul-tuli committed:
Updated signature + usages of `helpers.create_model`

rahul-tuli committed:
Refactor: `get_train_and_val_loaders` to `get_dataset_and_dataloader(...)`

Update: usages of the above function into two separate calls for training and validation datasets

Simplify: Logic, with credits to @bfineran

rahul-tuli committed:
Refactor: `get_train_and_val_loaders` to `get_dataset_and_dataloader(...)`

Update: usages of the above function into two separate calls for training and validation datasets

Simplify: Logic, with credits to @bfineran

rahul-tuli committed:
Fix: Create an empty context manager cause contextlib.nullcontext is only supported in  python 3.7+

rahul-tuli committed:
Fix: Create an empty context manager cause contextlib.nullcontext is only supported in  python 3.7+

rahul-tuli committed:
Update: Base branch
Address: Review comments from @dbogunowicz

rahul-tuli committed:
Address: Review comments from @bfineran

rahul-tuli committed:
Revert "fix get_default_mask_creator error message (#669)"

This reverts commit 507d2de7090672abcdda748f69b959e0e33cb03a.

rahul-tuli committed:
Revert "Revert "fix get_default_mask_creator error message (#669)""

This reverts commit 49071adb1f7358d1adfa3d084e7a7d23e15f32d5.

rahul-tuli committed:
Rebase on `feature/hepers-cleanup` and add support for the new trainer and `helpers.py` refactor

rahul-tuli committed:
Address: Review comments from @bfineran

* Move `ffcv_dataset.py` into it's own package

* minor nits

bfineran reviewed (commented):
let's also get support for `sparseml.pytorch.datasets.classification.ImageFolder`

bfineran reviewed (commented):
looking good overall. please update any comments that are now resolved, provide initial documentation on install, and list commands tested in the PR description

dbogunowicz reviewed (dismissed):
Looking good.

bfineran reviewed (changes requested):
blocking until we have instructions for install included

bfineran committed:
Merge branch 'main' into feature/helpers-cleanup

rahul-tuli committed:
Fix: Docstring commands
* A minor bug where lr and pr analysis scripts were still relying on the deprecrated get_loss_wrapper function from old helpers

rahul-tuli committed:
Merge branch 'main' into feature/helpers-cleanup

rahul-tuli committed:
Add: FFCV support to imagefolder.py
Add: README.md doc to install and use FFCV

bfineran commented on src/sparseml/pytorch/datasets/image_classification/README.md:
```diff
@@ -0,0 +1,35 @@
+
+### Installation
+
+Use conda to install all requirements; install conda using instructions from
```
```suggestion
Use conda to install all ffcv dependencies; install conda using instructions from
```

bfineran commented on src/sparseml/pytorch/datasets/image_classification/README.md:
```diff
@@ -0,0 +1,35 @@
+### FFCV SparseML support
```
let's rename the file to ffcv.md for now

bfineran commented on src/sparseml/pytorch/datasets/image_classification/README.md:
```diff
@@ -0,0 +1,35 @@
+
+### Usage
+
+To use [FFCV] with [SparseML], just pass in the `--ffcv` flag to training script.
```
```suggestion
To use [FFCV] with [SparseML], pass in the `--ffcv` flag to training script.
```

bfineran commented on src/sparseml/pytorch/datasets/image_classification/README.md:
```diff
@@ -0,0 +1,35 @@
+### FFCV SparseML support
+
+[FFCV] or Fast Forward Computer Vision is a drop-in data-loading system that 
```
nit: missing FFCV link

rahul-tuli replied:
[FFCV] anchor at the end

bfineran commented on src/sparseml/pytorch/datasets/image_classification/README.md:
```diff
@@ -0,0 +1,35 @@
+### FFCV SparseML support
+
+[FFCV] or Fast Forward Computer Vision is a drop-in data-loading system that 
+can dramatically increase data throughput in model training. [SparseML] now 
+provides experimental support for FFCV with it's [ImageNet] based datasets.
```
also missing links for sparseml and imagenet?

rahul-tuli replied:
Sparseml and Imagenet anchors at the end

bfineran reviewed (changes requested):
looks great @rahul-tuli. left comments on readme

rahul-tuli committed:
Add: Fixed DDP support for multi-gpu setup

rahul-tuli committed:
Renamed to ffcv.md

rahul-tuli committed:
Merge branch 'main' into ffcv-tests

rahul-tuli committed:
Updated docstring

rahul-tuli committed:
Fix Style
